### ISSUE TITLE
[Backend][powershell] Key name is incorrectly added into the regular expression

### ISSUE THREAD

**Issue by 0xballistics** [Neo23x0/sigma#762]
## Problem Definition:
Output of following command does not match related event:

### COMMAND:
```
sigmac --target powershell  -c tools/config/powershell.yml rules/windows/powershell/powershell_create_local_user.yml
```

### SIGMAC OUTPUT:
```
Get-WinEvent -LogName Microsoft-Windows-PowerShell/Operational | where {($_.ID -eq "4104" -and ($_.message -match ".*New-LocalUser.*")) } | select TimeCreated,Id,RecordId,ProcessId,MachineName,Message
```

### EVENT:
```
TimeCreated : 5/15/2020 6:03:50 PM
Id          : 4104
RecordId    : 100
ProcessId   : 5020
MachineName : DESKTOP-BSF8Q89
Message     : Creating Scriptblock text (1 of 1):
              New-LocalUser -Name "User06" -Description "Descriptasd" -NoPassword

              ScriptBlock ID: 852367ae-5db9-4583-9908-8ae1f809b951
              Path:
```

## Solution:
Problem can be solved replacing $\_.message -match "Message.\*New-LocalUser.\*" condition with $\_.message -match ".\*New-LocalUser.\*"

Cause of the problem seems to be incorrectly added key name at https://github.com/Neo23x0/sigma/blob/master/tools/sigma/backends/powershell.py#L142

## Explanation:
When we fetch the output of the related event log with the following command we can see there is no "Message" prefix at the return value
### COMMAND:
```
(Get-WinEvent -LogName Microsoft-Windows-PowerShell/Operational | where {($_.ID -eq "4104" -and ($_.message -match ".*New-LocalUser.*")) } | select TimeCreated,Id,RecordId,ProcessId,MachineName,Message).Message
```
### OUTPUT:
```
ScriptBlock ID: d4e0ffb8-b1c1-498a-9440-dd3158bf6f58
Path:
Creating Scriptblock text (1 of 1):
New-LocalUser -Name "User06" -Description "Descriptasd" -NoPassword
```

Tested with Powershell v5.1 and v7.0

**Comment by 0xballistics** [comment]
Fixed related lines:
https://github.com/0xballistics/sigma/commit/aa69b4020cbd8f556128ddf2c62ecb075e6932c1

I can do a PR but I have no idea if I am breaking any other detection logic.

**Comment by Karneades** [comment]
Hi @0xballistics 

I looked at your example but I think the backend is ok. Important for the discussion and for rule authors is to know, that for the PowerShell backend **we by design [match against the message field](https://github.com/Neo23x0/sigma/blob/master/tools/sigma/backends/powershell.py#L142), because only there is the real event data stored**, relevant line here:

=> itemslist.append('$_.**message** -match %s' % (self.generateValueNode(...))

That means PowerShell rules should only have the rules for matching what is inside the message field. Your [rule](https://github.com/Neo23x0/sigma/blob/master/rules/windows/powershell/powershell_create_local_user.yml#L20) looks like this:

```
detection:
    selection:
        EventID: 4104
        Message|contains:
            - 'New-LocalUser'
```

This advises the PowerShell backend to look inside the message field for a message field... which is unwished with the current implementation of the PowerShell backend which already matches against that field by default.

So, the message field in this rule is a duplicate. If a field is used in the rule, normally, one would like to match a that field inside the message field and not against the Windows standard event log fields like EventID, ...but inside the message field where the interesting data is found. For example in Sysmon where you then have various fields inside the message field, like the command line, parent image and you like to match only against them. For this purpose the key from the rule is used in the pattern.

**Comment by Karneades** [comment]
In summary to the above comment: The addition of the key field is on purpose and needed to be able to match against specific fields when using the PowerShell backend.

Unfortunately, the way Windows presents the events to PowerShell vs. when using log forwarding differs and is limited in the way we have access to the important event data (message field). 

I try to list some ways for mitigating these wrong conversions:

- **Check in the backend if the key name is _message_ and skip the addition of the key** name for that special name. This would **prevent any rule for matching against a _message_ field inside the _message_ field** when using the PowerShell backend but let's all other backends work where a field called "message" for the Windows event logs is available. I think this has the smallest impact of the whole thing, because I'm not aware of any Windows event log with a message field inside a message field... but you never know :) Important: using log forwarding for Windows event logs this generic field is called _EventData_ instead of _message_. So rules with a filter for _message_ won't work on some log systems, because there is no _message_ field at all.
- **Remove message field from all rules and rely on the implicit usage of the message field** when using the PowerShell backend (meeh, see https://github.com/Neo23x0/sigma/issues/501I) and for other backends match against the whole (raw) event without explicit field match.
- **Remove this implicit matching against message and only rely on rules having an explicit message field filter**.
    - **This makes matching against specific fields inside _message_ impossible for the PowerShell backend.**

Example Windows event which is seen by PowerShell in the following way, where we maybe would like to match against the whole message field:

```
TimeCreated : 5/15/2020 6:03:50 PM
Id          : 4104
RecordId    : 100
ProcessId   : 5020
MachineName : DESKTOP-BSF8Q89
Message     : Creating Scriptblock text (1 of 1):
              New-LocalUser -Name "User06" -Description "Descriptasd" -NoPassword

              ScriptBlock ID: 852367ae-5db9-4583-9908-8ae1f809b951
              Path:...
```

Example Windows event which is seen by PowerShell where we would like to be able to also match against specific field _inside_ the message field (Sysmon, security logs, PowerShell logs, ...):

```
TimeCreated : ...
Id          : 1
Message:
   Image: ..
   ParentImage..
   CommandLine...
   ...
```

[Example Windows event which is seen by other log systems when log forwarding](https://www.ultimatewindowssecurity.com/securitylog/encyclopedia/event.aspx?eventid=90011) is used. So there is no _message_ field: but a _EventData_:

```
 <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">

    <System>
        <Provider Name="Microsoft-Windows-Sysmon" Guid="{5770385F-C22A-43E0-BF4C-06F5698FFBD9}" />
        <EventID>11</EventID>
        <Version>2</Version>
        <Level>4</Level>
        <Task>11</Task>
        <Opcode>0</Opcode>
        <Keywords>0x8000000000000000</Keywords>
        <TimeCreated SystemTime="2017-05-13T19:44:55.314125100Z" />
        <EventRecordID>734181</EventRecordID>
        <Correlation />
        <Execution ProcessID="2848" ThreadID="3520" />
        <Channel>Microsoft-Windows-Sysmon/Operational</Channel>
        <Computer>rfsH.lab.local</Computer>
        <Security UserID="S-1-5-18" />
    </System>
    <EventData>
        <Data Name="UtcTime">2017-05-13 19:44:55.313</Data>
        <Data Name="ProcessGuid">{A23EAE89-6237-5917-0000-0010300E6601}</Data>
        <Data Name="ProcessId">19200</Data>
        <Data Name="Image">C:\Windows\Microsoft.NET\Framework64\v4.0.30319\mscorsvw.exe</Data>
        <Data Name="TargetFilename">C:\Windows\assembly\NativeImages_v4.0.30319_64\Temp\4b00-0\AxImp.exe</Data>
        <Data Name="CreationUtcTime">2017-05-13 19:44:55.313</Data>
    </EventData>
</Event>
```

**Comment by neu5ron** [comment]
@thomaspatzke SOC Prime will work on fix, since I have been in the elasticsearch backend code a lot recently. this was one of the remaining things I had recently discovered after trying to cleanup the mess of winlogbeat 7 and ECS removing analyzed fields completely and also changing .keyword completely

**Comment by 0xballistics** [comment]
@Karneades many thanks for the explanation. Now, I understand the problem.

So the best solution seems to be:

> Check in the backend if the key name is message and skip the addition of the key name.

and assume there won't be a nested message field :)

**Comment by Karneades** [comment]
And additionally, a user specific config is also needed when there is no _message_ field at all available (field mapping for _message_ to e.g. _EventData_). I think one of the seldom cases where there is a _message_ field is when using the PowerShell backend, many other log systems have the _EventData_ field based on the XML structure.

**Comment by neu5ron** [comment]
yup, this is beyond powershell as your seeing. this issue always existed even before updates because of elastic’s infinite schema/field name decisions. But it became a pretty glaring issue in 7.x with the drastic changes of ECS.

**Comment by Karneades** [comment]
> SOC Prime will work on fix, since I have been in the elasticsearch backend code a lot recently. this was one of the remaining things I had recently discovered after trying to cleanup the mess of winlogbeat 7 and ECS removing analyzed fields completely and also changing .keyword completely

> yup, this is beyond powershell as your seeing. this issue always existed even before updates 

@neu5ron Why do you bring ECS up while we discuss the issue for the PowerShell backend? Which part of the issue is ECS related? When you say _SOC Prime will work on fix_, what fix are you talking about?

The thing with the #501 is of course beyond PowerShell (only have rules bound to fields), but the issue with the matching-approach is PowerShell backend specific (implicit matching for the message field).

**Comment by neu5ron** [comment]
apologies for the confusion, I was having two conversations that are same issue - of the bound fields and conditions that contain `keywords`

**Comment by 0xballistics** [comment]
Hi guys! Any updates on this one? @neu5ron did you have chance to look into it? 

2b72ee7b847fe96919a0b13682171f87031e6379 must be providing an acceptable fix if there are no issues besides the concerns @Karneades stated.

**Comment by Karneades** [comment]
This commit fixes the issue with the PowerShell backend and when rules have a message filter. 

For PowerShell rules in general we have to handle also the following: Last year many PowerShell rules were given a message filter because some backend systems require a field-bounded search. All rules with a message filter would currently fail on all other log systems because there is **no** message field at all and a transformation from message to the log system specific word (e.g. EventData) is missing in backend config.
- [ ] Add new issue that all PowerShell rules with a message filter fail. Either use a config to transform message to the correct form or handle it otherweise. Removing the generic message filter would solve that issue to but would be against #501 which was about to add field-specific filters for all rules.

**Comment by thomaspatzke** [comment]
> 2b72ee7b847fe96919a0b13682171f87031e6379 must be providing an acceptable fix if there are no issues besides the concerns @Karneades stated.

Can you provide it as pull request? Then we could integrate it to improve the situation.

@Karneades as far as I understood the issue the best solution would be to match the message field to something appropriate in the used backends config?

**Comment by 0xballistics** [comment]
Sure! Made the pull request

**Comment by Karneades** [comment]
> as far as I understood the issue the best solution would be to match the message field to something appropriate in the used backends config?

@thomaspatzke Yep, the field name from the Windows event is `EventData`, in my case we use `EventData_Xml` as the corresponding "message" field. But now everyone must use "message" as the field name in rules for a generic message filter so that the PowerShell backend works now as expected.
